projects/online/online/utils/buffer.py

```diff
@@ -0,0 +1,163 @@
+import h5py
+import numpy as np
+import torch
+
+
+class InputBuffer:
+    """
+    A buffer for storing raw strain data f
+    or use in parameter estimation (amplfi) followup
+    of events detected by Aframe
+
+    Args:
+        num_channels:
+            The number of channels in the data
+        sample_rate:
+            The sampling rate of the data
+        buffer_length:
+            The length of the buffer in seconds
+        pe_window:
+            The length of the window to use for parameter estimation in seconds
+        event_position:
+            The placement of the coalescence time
+            of the event in the pe window in seconds
+    """
+
+    def __init__(
+        self,
+        num_channels: int,
+        sample_rate: float,
+        buffer_length: float,
+        fduration: float,
+        pe_window: float,
+        event_position: float,
+    ):
+        self.fduration = fduration
+        self.num_channels = num_channels
+        self.sample_rate = sample_rate
+        self.buffer_length = buffer_length
+        self.buffer_size = int(buffer_length * sample_rate)
+        self.pe_window = pe_window
+        self.event_position = event_position
+        self.reset()
+
+    def write(self, write_path, event_time):
+        start = self.t0
+        stop = self.t0 + self.buffer_length
+        time = np.linspace(start, stop, self.buffer_size)
+        with h5py.File(write_path, "w") as f:
+            f.attrs.create("event_time", data=event_time)
+            f.create_dataset("time", data=time)
+            f.create_dataset("H1", data=self.input_buffer[0, :].cpu())
+            f.create_dataset("L1", data=self.input_buffer[1, :].cpu())
+
+    def reset(self):
+        self.t0 = None
+        self.input_buffer = torch.zeros(
+            (self.num_channels, self.buffer_size), device="cuda"
+        )
+
+    def update(self, update, t0):
+        self.input_buffer = torch.cat([self.input_buffer, update], axis=-1)
+        self.input_buffer = self.input_buffer[:, -self.buffer_size :]
+        update_duration = update.shape[-1] / self.sample_rate
+        self.t0 = t0 - (self.buffer_length - update_duration)
+
+    def get_pe_data(self, event_time: float):
```

Comment:
It feels a little weird to have this method be part of the input buffer, but I think it's as good as any alternative

Reply:
I added it here since it seemed that the `InputBuffer` only served to extract data for the PE model.

Reply:
Yeah, I think it makes sense. I can imagine down the line that we might want to use this data for other purposes, but not at the moment.